Add php8 support and remove obsolete versions

codisart opened the pull request:
Issue #6 

Q | A
-- | --
New Feature | yes


- [x] Modify composer.json to provide support for PHP 8.0 by adding the constraint ~8.0.0
- [x] Modify composer.json to drop support for PHP less than 7.3
- [x] Modify composer.json to implement phpunit 9.3 which supports PHP 7.3+
- [x] Modify .travis.yml to ignore platform requirements when installing composer dependencies (simply add --ignore-platform-reqs to COMPOSER_ARGS env variable)
- [x] Modify .travis.yml to add PHP 8.0 to the matrix (NOTE: Do not allow failures as PHP 8.0 has a feature freeze since 2020-08-04!)
- [x] Modify source code in case there are incompatibilities with PHP 8.0

codisart committed:
Add php8 support and remove obsolete versions

Signed-off-by: codisart <louis.celier@gmail.com>

codisart committed:
Fix phpunit 9 deprecations

Signed-off-by: codisart <louis.celier@gmail.com>

codisart committed:
Fix php8 compatibilities

Signed-off-by: codisart <louis.celier@gmail.com>

michalbundyra commented on src/ClassFileLocator.php:
```diff
@@ -112,15 +112,17 @@ public function accept()
                             case T_STRING:
                             case T_NS_SEPARATOR:
+                            case 312: // T_NAME_FULLY_QUALIFIED
+                            case 314: // T_NAME_QUALIFIED
```
this is not gonna work, these constant can have different values on different PHP version, see for example:
https://3v4l.org/XFk83

I think, what we need here, is to define these constants if not defined - but with string values, something what has been done in PHP_CodeSniffer. Then it's  gonna be just placeholder on older PHP versions and should behave as expects on PHP 8.0+

I think to do that, we would need separate autoload.php file, add it to composer and define the constants there.

codisart replied:
Hi @michalbundyra,

You're right but I'm not sure to understand fully what you propose. 
Do you have an example to help me implement your proposition ?

boesing replied:
@codisart Not sure if I get this right, but if we need those tokens on PHP 8.0+ only, you could add something like:

```php

if (PHP_VERSION_ID > 80000) {
    switch ($type) {
        case T_NAME_FULLY_QUALIFIED:
        case T_NAME_QUALIFIED:
            $namespace .= $content;
         break;
    }
}
```

This would use the constants when PHP version is greather than PHP 8.0.0 🤷‍♂️

codisart replied:
I see that php cs used the same method https://github.com/squizlabs/PHP_CodeSniffer/pull/3063/files

codisart replied:
I really hope I am not creating a vicious technical debt. ^^

boesing replied:
As there is no other way to properly support this, I think there is no other way.
Even tho, there should be already a unit test for covering this. 🤷🏼‍♂️

codisart replied:
The modification is done

michalbundyra reviewed (commented):
@codisart thanks for you contribution. In general it looks good, just one comment around token constants.

Thanks!

boesing commented on src/ClassFileLocator.php:
```diff
@@ -112,15 +112,16 @@ public function accept()
+                        if (PHP_VERSION_ID >= 80000) {
+                            $types = [T_NAME_FULLY_QUALIFIED, T_NAME_QUALIFIED, T_STRING, T_NS_SEPARATOR];
+                        }
+                        if (in_array($type, $types)) {
```
```suggestion
                        if (in_array($type, $types, true)) {
```

Prefering strict type comparison is always recommended.

boesing commented on src/ClassFileLocator.php:
```diff
@@ -112,15 +112,16 @@ public function accept()
+                        list($type, $content) = $token;
+                        $types = [T_STRING, T_NS_SEPARATOR];
+                        if (PHP_VERSION_ID >= 80000) {
+                            $types = [T_NAME_FULLY_QUALIFIED, T_NAME_QUALIFIED, T_STRING, T_NS_SEPARATOR];
```
I would prefer having `array_merge` over re-assigning the all values as its easier to understand that PHP 8 just adds two new constants.

```suggestion
                            $types = array_merge($types, [T_NAME_FULLY_QUALIFIED, T_NAME_QUALIFIED]);
```

codisart replied:
that is a better way.

boesing commented on src/ClassFileLocator.php:
```diff
@@ -112,15 +112,16 @@ public function accept()
                         }
                     }
-                    if ($saveNamespace) {
+                    if (isset($saveNamespace) && $saveNamespace) {
```
Please remove the `isset` and ensure that `saveNamespace` is always available as `false`.

boesing reviewed (changes requested):
Thanks for applying the latest changes.
I've found 2 more things which are introduced with these changes.

One thing I haven't seen in the reviews before was that `isset($saveNamespace)`. Please get rid of that by creating that variable when the iteration starts. This will ensure it is always available and thus doesn't need the `isset` check.

weierophinney committed:
qa: update lockfile from PHP 7.4 and ignore phpunit result cache

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: remove Prophecy usage

Use native mocks instead.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: fix test expectation

PHPUnit has changed how "assertEquals" works with arrays, and now order matters for some reason.
Re-ordered expectations.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>
